torch/csrc/autograd/compiled_autograd.h

```diff
@@ -0,0 +1,583 @@
+    auto fn_id = _compiler.emplace_hook(std::move(obj));
+    collect_size(fn_id);
+    _node_call.post_hooks.emplace_back(fn_id);
+  }
+
+  void collect_size(size_t s) {
+    // we expect sizes to be small, so try to cram them into a single byte
+    // shorter keys, means faster caches
+    constexpr uint8_t encode_as_u64 = std::numeric_limits<uint8_t>::max();
+    constexpr uint8_t encode_as_u32 = encode_as_u64 - 1;
+    constexpr uint8_t encode_as_u16 = encode_as_u64 - 2;
+    if (C10_UNLIKELY(s >= encode_as_u16)) {
+      // first write a byte indicating the path we followed, then the data
+      if (s <= std::numeric_limits<uint16_t>::max()) {
+        // 3 bytes
+        specialize_on_bytes(encode_as_u16);
+        specialize_on_bytes(static_cast<uint16_t>(s));
+      } else if (s <= std::numeric_limits<uint32_t>::max()) {
+        // 5 bytes
+        specialize_on_bytes(encode_as_u32);
+        specialize_on_bytes(static_cast<uint32_t>(s));
+      } else {
+        // 9 bytes
+        specialize_on_bytes(encode_as_u64);
+        specialize_on_bytes(s);
+      }
+    } else {
+      // happy case, 1 byte
+      specialize_on_bytes(static_cast<uint8_t>(s));
+    }
+  }
+
+  SizeInput::DynType set_default_dyn_type(SizeInput::DynType default_dyn_type) {
+    return std::exchange(_compiler.default_dyn_type, default_dyn_type);
+  }
+
+  CompiledNodeArgs(AutogradCompilerCall& compiler, NodeCall& node_call)
+      : _compiler(compiler),
+        _node_call(node_call),
+        _specialization_key_size(0),
+        _specialization_key_storage(1024),
+        _specialization_key(
+            (uint8_t*)std::malloc(_specialization_key_storage)) {}
+  ~CompiledNodeArgs() {
+    std::free(_specialization_key);
+  }
+  CompiledNodeArgs(const CompiledNodeArgs&) = delete;
+
+ private:
+  template <typename T>
+  void specialize_on_bytes(const T& t) {
+    while (C10_UNLIKELY(
+        _specialization_key_size + sizeof(T) > _specialization_key_storage)) {
+      _specialization_key_storage *= 2;
+      _specialization_key = (uint8_t*)std::realloc(
+          _specialization_key, _specialization_key_storage);
+    }
+    std::memcpy(_specialization_key + _specialization_key_size, &t, sizeof(T));
+    _specialization_key_size += sizeof(T);
+  }
+
+  AutogradCompilerCall& _compiler;
+  NodeCall& _node_call;
+  size_t _specialization_key_size;
+  size_t _specialization_key_storage;
+  uint8_t* _specialization_key;
+};
+
+struct TraceState {
+  TraceState(
+      const variable_list& pi,
+      const std::vector<c10::optional<c10::SymInt>>& ss,
+      bool accumulate_grad_,
+      size_t num_outputs)
+      : proxy_inputs_index(0),
+        sym_sizes_index(0),
+        proxy_inputs(pi),
+        sym_sizes(ss),
+        outputs(num_outputs),
+        accumulate_grad(accumulate_grad_) {}
+
+  ~TraceState() {
+    if (C10_UNLIKELY(proxy_inputs_index != proxy_inputs.size())) {
+      TORCH_WARN("not all proxy_inputs consumed")
```

Comment:
can we manually destruct TraceState via a non destructor method, and the turn these into errors?